applications/ConvectionDiffusionApplication/custom_utilities/embedded_mls_constraint_process.cpp

```diff
@@ -0,0 +1,437 @@
+                    Matrix cloud_nodes_coordinates;
+                    PointerVector<NodeType> cloud_nodes;
+                    SetNegativeNodeSupportCloud(r_node, cloud_nodes, cloud_nodes_coordinates);
+
+                    // Calculate the MLS basis in the current negative node
+                    Vector N_container;
+                    const array_1d<double,3> r_coords = r_node.Coordinates();
+                    const double mls_kernel_rad = CalculateKernelRadius(cloud_nodes_coordinates, r_coords);
+                    p_mls_sh_func(cloud_nodes_coordinates, r_coords, 1.01 * mls_kernel_rad, N_container);
+
+                    // Save the extension operator nodal data
+                    std::size_t n_cl_nod = cloud_nodes.size();
+                    CloudDataVectorType cloud_data_vector(n_cl_nod);
+                    for (std::size_t i_cl_nod = 0; i_cl_nod < n_cl_nod; ++i_cl_nod) {
+                        auto p_cl_node = cloud_nodes(i_cl_nod);
+                        auto i_data = std::make_pair(p_cl_node, N_container[i_cl_nod]);
+                        cloud_data_vector(i_cl_nod) = i_data;
+                    }
+
+                    auto ext_op_key_data = std::make_pair(&r_node, cloud_data_vector);
+                    rExtensionOperatorMap.insert(ext_op_key_data);
+                }
+            }
+        }
+    }
+}
+
+void EmbeddedMLSConstraintProcess::ApplyExtensionConstraints(NodesCloudMapType& rExtensionOperatorMap)
+{
+    // Initialize counter of master slave constraints
+    ModelPart::IndexType id = mpModelPart->NumberOfMasterSlaveConstraints()+1;
+    // Get variable to constrain 
+    const auto& r_var = KratosComponents<Variable<double>>::Get("TEMPERATURE");
+
+    // Loop through all negative nodes of split elements (slave nodes)
+    for (auto it_slave = rExtensionOperatorMap.begin(); it_slave != rExtensionOperatorMap.end(); ++it_slave) {
+        auto p_slave_node = std::get<0>(*it_slave);
+        auto& r_ext_op_data = rExtensionOperatorMap[p_slave_node];
+
+        // Add one master slave constraint for every node of the support cloud (master) of the negative node (slave)
+        // The contributions of each master will be summed up in the BuilderAndSolver to give an equation for the slave dof
+        for (auto it_data = r_ext_op_data.begin(); it_data != r_ext_op_data.end(); ++it_data) {
+            auto& r_node_data = *it_data;
+            auto p_support_node = std::get<0>(r_node_data);
+            const double support_node_N = std::get<1>(r_node_data);
+
+            // Add master slave constraint, the support node MLS shape function value N serves as weight of the constraint
+            mpModelPart->CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", id++, 
+            *p_support_node, r_var, *p_slave_node, r_var, 
+            support_node_N, 0.0);
+        }
+    }  
+}
+
+void EmbeddedMLSConstraintProcess::SetInterfaceFlags()

```

Comment:
Are you sure you need all these to be set for this application? If not I'd simplify as much as possible.

Reply:
Well, those flags are used in the `SetNegativeNodeSupportCloud` method that you implemented and it makes it easier to distinguish between nodes of negative elements and nodes of intersected elements. It could also be done by only using the ACTIVE flag.

This also raises the question whether we want to be able to visualize the BOUNDARY and INTERFACE in the post-processing. We could skip the entire search for neighboring elements and only mark the nodes.